compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -2237,7 +2237,10 @@ trait Applications extends Compatibility {
 
         def isCorrectUnaryFunction(alt: TermRef): Boolean =
           val formals = params(alt)
-          formals.length == 1 && ptIsCorrectProduct(formals.head, args)
+          formals.lengthIs == 1 && {
+            ptIsCorrectProduct(formals.head, args)
+            | formals.head.dealias.isInstanceOf[TypeParamRef] // eg. i21682
```

Comment:
I think we need an explanation what goes on here and why the test is necessary and sufficient. Right now I have no idea what goes on here.